Certificate column empty on all instances

The certificate column is empty on all listed instances on searx.space


<img width="1439" alt="searx space 2022-01-02 21 39 05" src="https://user-images.githubusercontent.com/77782671/147888989-9b9893dd-7968-474d-bdcb-28881ddd8c74.png">

From https://github.com/searxng/searx-space/commit/1b618f2ce920d27afc3db0741487c48270eca374#r64431924 ( @not-my-profile ):

https://github.com/searxng/searx-space/blob/1b618f2ce920d27afc3db0741487c48270eca374/searxstats/common/ssl_info.py#L5

> `load_certificate` and `FILETYPE_ASN1` are still used in `update_obj_with_bin` ... which is also the cause of the CI failure:
> 
> ```
> searxstats/common/ssl_info.py:29:15: F821 undefined name 'load_certificate'
> searxstats/common/ssl_info.py:29:32: F821 undefined name 'FILETYPE_ASN1'
> ```
> 
> Is `update_obj_with_bin` never used or why doesn't this matter?

--- a/Makefile
+++ b/Makefile
@@ -1,10 +1,6 @@
 APP_NAME=searx/searxstats:latest
-CONTAINER_NAME=searxstats
-OUTPUT_FILENAME=instances.json
 
 ROOT_DIR:=$(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
-
-DOCKER_INTERACTIVE_PARAM:=$(shell [ -t 0 ] && echo " -t -i")
 
 qa:
 	flake8 --max-line-length=120 searxstats tests
@@ -25,18 +21,7 @@ docker-run: # Run the container
 	chgrp 1005 $(ROOT_DIR)/cache
 	chmod 775 $(ROOT_DIR)/cache
 	# run
-	docker run $(DOCKER_INTERACTIVE_PARAM) \
-	    --rm \
-	    --network host \
-	    -v $(ROOT_DIR)/html/data:/usr/local/searxstats/html/data \
-	    -v $(ROOT_DIR)/cache:/usr/local/searxstats/cache \
-	    -e MMDB_FILENAME=/usr/local/searxstats/cache/dbip-country.mmdb \
-	    --name="$(CONTAINER_NAME)" \
-	    $(APP_NAME) \
-	    --database sqlite:////usr/local/searxstats/cache/searxstats.db \
-	    --cache /usr/local/searxstats/cache \
-	    --output /usr/local/searxstats/html/data/$(OUTPUT_FILENAME) \
-	    --all
+	./docker-run.sh --all
 
 webserver:
 	cd $(ROOT_DIR)/html; python -m http.server 8889

--- a/docker-run.sh
+++ b/docker-run.sh
@@ -1,0 +1,18 @@
+ROOT_DIR=$(dirname $(realpath $0))
+CONTAINER_NAME=searxstats
+DOCKER_INTERACTIVE_PARAM=$([ -t 0 ] && echo " -t -i")
+APP_NAME=searx/searxstats:latest
+OUTPUT_FILENAME=instances.json
+
+docker run $DOCKER_INTERACTIVE_PARAM \
+    --rm \
+    --network host \
+    -v $ROOT_DIR/html/data:/usr/local/searxstats/html/data \
+    -v $ROOT_DIR/cache:/usr/local/searxstats/cache \
+    -e MMDB_FILENAME=/usr/local/searxstats/cache/dbip-country.mmdb \
+    --name="$CONTAINER_NAME" \
+    $APP_NAME \
+    --database sqlite:////usr/local/searxstats/cache/searxstats.db \
+    --cache /usr/local/searxstats/cache \
+    --output /usr/local/searxstats/html/data/$OUTPUT_FILENAME \
+    "$@"

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,4 +1,4 @@
-pyOpenSSL==19.1.0
+pyOpenSSL==21.0.0
 httpx==0.11.0
 brotlipy==0.7.0
 uvloop==0.14.0

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,3 +1,4 @@
+pyOpenSSL==19.1.0
 httpx==0.11.0
 brotlipy==0.7.0
 uvloop==0.14.0

--- a/searxstats/__main__.py
+++ b/searxstats/__main__.py
@@ -1,5 +1,7 @@
 import argparse
 import asyncio
+import os
+import sys
 
 from .common.memoize import bind_to_file_name
 from .config import (
@@ -13,6 +15,13 @@ from . import initialize, run_once, run_server, erase_memoize, finalize
 # pylint: disable=too-many-locals
 def run(server_mode: bool, output_file_name: str, user_cache_directory: str, database_url: str,
         instance_urls: list, private: bool, selected_fetcher_names: list, update_fetcher_memoize_list: list):
+
+    if not os.access(user_cache_directory, os.W_OK):
+        sys.exit('[FATAL ERROR] need write access to {}'.format(user_cache_directory))
+
+    if not os.access(output_file_name, os.W_OK):
+        sys.exit('[FATAL ERROR] need write access to {}'.format(output_file_name))
+
     private_str = 'Private' if private else 'Public'
     server_mode_str = 'server mode' if server_mode else 'single run'
     server_emoji_str = '🤖' if server_mode else '⚡'

--- a/searxstats/common/http.py
+++ b/searxstats/common/http.py
@@ -12,7 +12,7 @@ import httpx.backends.asyncio
 from .utils import exception_to_str
 from .queuecalls import UseQueue
 from .memoize import Memoize
-from .ssl_info import get_httpx_backend
+from .ssl_info import SSL_CONTEXT
 from ..config import TOR_HTTP_PROXY
 
 if not sys.version_info.major == 3 and sys.version_info.minor >= 7:
@@ -66,9 +66,7 @@ async def new_client(*args, **kwargs):
             network_type = kwargs['network_type']
             kwargs['proxies'] = NETWORK_PROXIES.get(network_type, None)
         del kwargs['network_type']
-    if 'backend' not in kwargs:
-        kwargs['backend'] = get_httpx_backend()
-    async with httpx.AsyncClient(*args, **kwargs) as session:
+    async with httpx.AsyncClient(*args, backend='asyncio', verify=SSL_CONTEXT) as session:
         session._network_type = network_type  # pylint: disable=protected-access
         yield session
 

--- a/searxstats/common/ssl_info.py
+++ b/searxstats/common/ssl_info.py
@@ -2,7 +2,7 @@ import ssl
 import httpx
 import httpx.config
 import httpx.backends.asyncio
-# from OpenSSL.crypto import load_certificate, FILETYPE_ASN1
+from OpenSSL.crypto import load_certificate, FILETYPE_ASN1
 
 
 def set_or_concat_value(obj, key, value):

--- a/searxstats/common/ssl_info.py
+++ b/searxstats/common/ssl_info.py
@@ -1,7 +1,5 @@
 import ssl
-import httpx
-import httpx.config
-import httpx.backends.asyncio
+from typing import Dict
 from OpenSSL.crypto import load_certificate, FILETYPE_ASN1
 
 
@@ -47,56 +45,38 @@ def update_obj_with_bin(cert_obj, cert_bin):
             cert_obj['subject']['altName'] = str(ex)
 
 
-class SslInfo:
+SSL_CONTEXT = ssl.create_default_context()
 
-    __slots__ = ['_ssl_info']
+_SSL_OBJECTS: Dict[str, ssl.SSLObject] = {}
 
-    def __init__(self):
-        self._ssl_info = dict()
-
-    def parse_sslobject(self, hostname: str, sslobj: ssl.SSLObject):
-        if sslobj is None:
-            return
-        if hostname not in self._ssl_info:
-            cert_dict = sslobj.getpeercert(binary_form=False)
-            cert_bin = sslobj.getpeercert(binary_form=True)
-            # make cert_obj using cert_dict and cert_bin
-            cert_obj = cert_to_obj(cert_dict)
-            if cert_bin is not None and 'sha256' not in cert_obj:
-                update_obj_with_bin(cert_obj, cert_bin)
-            # store values
-            self._ssl_info[hostname] = {
-                'version': sslobj.version(),
-                'certificate': cert_obj
-            }
-
-    def get(self, hostname: str):
-        return self._ssl_info.get(hostname, {})
+_wrap_bio = SSL_CONTEXT.wrap_bio
 
 
-class AsyncioBackendLogCert(httpx.backends.asyncio.AsyncioBackend):
-
-    __slots__ = ['_sslinfo']
-
-    def __init__(self, sslinfo: SslInfo):
-        super().__init__()
-        self._sslinfo = sslinfo
-
-    async def open_tcp_stream(self, hostname, port, ssl_context, timeout):
-        value = await super().open_tcp_stream(hostname, port, ssl_context, timeout)
-        sslobj = value.stream_reader._transport.get_extra_info('ssl_object')  # pylint: disable=protected-access
-        self._sslinfo.parse_sslobject(hostname, sslobj)
-        return value
+def patched_wrap_bio(incoming: ssl.MemoryBIO, outgoing: ssl.MemoryBIO, server_hostname: str, **kwargs) -> ssl.SSLObject:
+    global _SSL_OBJECTS  # pylint: disable=global-statement
+    ssl_object = _wrap_bio(incoming, outgoing, server_hostname=server_hostname, **kwargs)
+    _SSL_OBJECTS[server_hostname] = ssl_object
+    return ssl_object
 
 
-SSLINFO = SslInfo()
-
-
-def get_httpx_backend():
-    global SSLINFO  # pylint: disable=global-statement
-    return httpx.backends.asyncio.AsyncioBackend()
+# we monkey patch SSL_CONTEXT to store SSLObjects in _ssl_objects
+# (subclassing ssl.SSLContext for some reason didn't work reliably)
+SSL_CONTEXT.wrap_bio = patched_wrap_bio
 
 
 def get_ssl_info(hostname):
-    global SSLINFO  # pylint: disable=global-statement
-    return SSLINFO.get(hostname)
+    global _SSL_OBJECTS  # pylint: disable=global-statement
+    ssl_object = _SSL_OBJECTS.get(hostname)
+    if ssl_object:
+        cert_dict = ssl_object.getpeercert(binary_form=False)
+        cert_bin = ssl_object.getpeercert(binary_form=True)
+        # make cert_obj using cert_dict and cert_bin
+        cert_obj = cert_to_obj(cert_dict)
+        if cert_bin is not None and 'sha256' not in cert_obj:
+            update_obj_with_bin(cert_obj, cert_bin)
+        return {
+            'version': ssl_object.version(),
+            'certificate': cert_obj
+        }
+    else:
+        return {}

